File: app/src/main/java/tech/pauly/findapet/data/ObservableHelper.java
package tech.pauly.findapet.data;

import javax.inject.Inject;

import io.reactivex.ObservableTransformer;
import io.reactivex.Scheduler;
import io.reactivex.SingleTransformer;
import tech.pauly.findapet.dependencyinjection.IoScheduler;
import tech.pauly.findapet.dependencyinjection.MainThreadScheduler;

public class ObservableHelper {
    private final Scheduler ioScheduler;
    private final Scheduler mainThreadScheduler;

    @Inject
    public ObservableHelper(@IoScheduler Scheduler ioScheduler, @MainThreadScheduler Scheduler mainThreadScheduler) {
        this.ioScheduler = ioScheduler;
        this.mainThreadScheduler = mainThreadScheduler;
    }

    public <T> ObservableTransformer<T, T> applyObservableSchedulers() {
        return upstream -> upstream.subscribeOn(ioScheduler)
                                   .observeOn(mainThreadScheduler);
    }

    public <T> SingleTransformer<T, T> applySingleSchedulers() {
        return upstream -> upstream.subscribeOn(ioScheduler)
                                   .observeOn(mainThreadScheduler);
    }
}

File: app/src/main/java/tech/pauly/findapet/data/ObservableHelper.kt
package tech.pauly.findapet.data

import io.reactivex.ObservableTransformer
import io.reactivex.Scheduler
import io.reactivex.SingleTransformer
import tech.pauly.findapet.dependencyinjection.IoScheduler
import tech.pauly.findapet.dependencyinjection.MainThreadScheduler
import javax.inject.Inject

open class ObservableHelper @Inject
constructor(@IoScheduler private val ioScheduler: Scheduler,
            @MainThreadScheduler private val mainThreadScheduler: Scheduler) {

    open fun <T> applyObservableSchedulers(): ObservableTransformer<T, T> {
        return ObservableTransformer {
            it.subscribeOn(ioScheduler).observeOn(mainThreadScheduler)
        }
    }

    open fun <T> applySingleSchedulers(): SingleTransformer<T, T> {
        return SingleTransformer {
            it.subscribeOn(ioScheduler).observeOn(mainThreadScheduler)
        }
    }
}
